modem: modem_cellular: add U-blox SARA-R5 as modem

emillindq opened the pull request:
Some modems, for example U-blox SARA-R5 autostarts and will power off
on the intended power on pulse. This introduces a dts bool flag to
signify it should skip power on pulse.

rerickson1 reviewed (changes requested):
If this is now optional, the driver will likely need changes to deal with the case where this pin is not defined.

emillindq commented:
> We can also be a bit proactive and add compatibility to the modem_cellular.c driver, a flag like `bool autostarts`, or a better name than that, which makes the driver skip the power pin on resume :)

Yeah, I realized it's probably a bad idea to just cut the power. Even in the case of autostart, it should power it off gracefully, otherwise it won't remember the last network and scan will take a long time next power on. The problem though is that it should still start suspended, and resume on device resume.

That can be managed with using the power enable as my "reset pin" though, and future users should route the reset pin from the sara-r5 (which we didn't do...) to manage that.

I'll cook something together!

emillindq commented:
So if I understand the vision correctly, it's basically to replace all individual modem drivers with modem_cellular.c.

In this commit now I've created a zephyr,modem-common.yaml which for now only contains the autostart boolean. According to the vision then, it could make sense to add the power key pin and reset pin there as well (as not required) since modem_cellular.c expects to operate power on/off with these two pins. What do you think @bjarki-trackunit? (BTW excellent work with the modem cellular, a huge improvement over offloaded individual drivers!!!).

bjarki-trackunit commented:
The modem_cellular driver is an MVP driver, which sacrifices modem model specific optimizations and features for the bare minimum "just works" features :)

The modem bindings files are not generic, they are specific to each modem, and used by other drivers which are optimized for that particular modem model.

The R5 requires its own binding, so the drivers can tell it apart from other modems. The modem cellular driver is then expanded to include said compatible.

It looks a bit misleading as many bindings reuse the same names for the power key and reset pins, but this is just for convinience, the modem_cellular driver can be adjusted to pick the power key even if the name is different for a paticular modem, see the many DEVICE_DEFINE macros in the modem_cellular driver :)

You only need to add the autostarts to the modem_cellular driver, then add an entry for the R5, and and set the flag in its DEVICE_DEFINE macro :)

emillindq committed:
modem: modem_cellular: add U-blox SARA-R5 as modem

U-blox SARA-R4 already exists but the behavior is different,
requiring a separate driver instance. For instance, R5 autostarts,
so this commit also adds support for skipping power on pulse.

Signed-off-by: Emil Lindqvist <emil@lindq.gr>

bjarki-trackunit commented on drivers/modem/modem_cellular.c:
```diff
@@ -444,7 +445,7 @@ static void modem_cellular_idle_event_handler(struct modem_cellular_data *data,
 
 	switch (evt) {
 	case MODEM_CELLULAR_EVENT_RESUME:
-		if (modem_cellular_gpio_is_enabled(&config->power_gpio)) {
+		if (modem_cellular_gpio_is_enabled(&config->power_gpio) && !config->autostarts) {
 			modem_cellular_enter_state(data, MODEM_CELLULAR_STATE_POWER_ON_PULSE);
 			break;
 		}
```
Could you change this to
```
	case MODEM_CELLULAR_EVENT_RESUME:
		if (config->autostarts) {
			modem_cellular_enter_state(data, MODEM_CELLULAR_STATE_AWAIT_POWER_ON);
			break;
		}

		if (modem_cellular_gpio_is_enabled(&config->power_gpio)) {
			modem_cellular_enter_state(data, MODEM_CELLULAR_STATE_POWER_ON_PULSE);
			break;
		}

		...
```

Without this change, the init script is sent immediately, which will not work for modems which require time to start up before they accept commands.

If the sara R5 does not need any time before accepting commands (it should probably at least get a few seconds after power on), you can adjust the startup time down for the sara R5 in the `MODEM_CELLULAR_DEVICE_U_BLOX_SARA_R5`
```
		.startup_time_ms = 10000,                                                          \
```

bjarki-andreasen reviewed (approved):
Nice work :)
